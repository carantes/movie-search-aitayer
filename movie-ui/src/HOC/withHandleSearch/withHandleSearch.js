import React, { Component } from "react";
import PropTypes from "prop-types";

const withHandleSearch = WrappedComponent => {
  class With extends Component {
    constructor(props) {
      super(props);
      this.timer = null;

      this.handleChange = this.handleChange.bind(this);
    }

    handleChange = e => {
      clearTimeout(this.timer);
      const param = e.target.value;
      const { onSearch, startWith = 3, waitInterval = 0 } = this.props;
      this.timer = setTimeout(() => {
        if (param.length >= startWith) {
          onSearch(param);
        }
      }, waitInterval);
    };

    render() {
      return <WrappedComponent onChange={this.handleChange} {...this.props} />;
    }
  }

  // With.defaultProps = {
  //   startWith: 3,
  //   waitInterval: 0,
  // };

  // With.propTypes = {
  //   onSearch: PropTypes.func.isRequired,
  //   startWith: PropTypes.number,
  //   waitInterval: PropTypes.number,
  // };

  return With;
};

export default withHandleSearch;
