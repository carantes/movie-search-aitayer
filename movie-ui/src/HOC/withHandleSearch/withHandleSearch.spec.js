import React from "react";
import { mount } from "enzyme";
import withHandleSearch from "./withHandleSearch";

describe("Search", () => {
  const waittime = 1;
  let stub = jest.fn();
  let search;

  beforeEach(() => {
    const HOC = withHandleSearch(props => <input {...props} />);
    search = mount(<HOC onSearch={stub} waitInterval={waittime} />);
  });

  it("should render properly", () => {
    expect(search).toMatchSnapshot();
  });

  it("should do nothing before type 3 characters", done => {
    const input = search.find("input").at(0);
    input.instance().value = "ah";
    input.simulate("change");

    setTimeout(() => {
      expect(stub.mock.calls.length).toBe(0);
      done();
    }, waittime);
  });

  it("should callback search function after type 3 characters", done => {
    const input = search.find("input").at(0);
    input.instance().value = "foo";
    input.simulate("change");

    setTimeout(() => {
      expect(stub).toHaveBeenCalled();
      done();
    }, waittime);
  });

  it("should respect waittime", function() {
    jest.useFakeTimers();
    search.find("input").simulate("change", { target: { value: "Foo" } });
    expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), waittime);
    jest.useRealTimers();
  });
});
