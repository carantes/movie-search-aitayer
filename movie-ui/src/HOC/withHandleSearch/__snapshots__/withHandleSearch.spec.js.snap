// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should render properly 1`] = `
<With
  onSearch={[MockFunction]}
  startWith={3}
  waitInterval={1}
>
  <Component
    onChange={[Function]}
    onSearch={[MockFunction]}
    startWith={3}
    waitInterval={1}
  >
    <input
      onChange={[Function]}
      onSearch={[MockFunction]}
      startWith={3}
      waitInterval={1}
    />
  </Component>
</With>
`;
